js/logic.js

```diff
@@ -0,0 +1,862 @@
+    this.maxDepth = opts.maxDepth || 10;
+    this.maxStringLength = opts.maxStringLength || 1000;
+    this.maxArrayLength = opts.maxArrayLength || 1000;
+    this.maxObjectLength = opts.maxObjectLength || 10;
+  }
+
+  ObjectSimplifier.prototype = {
+    simplify: function(x) {
+      return this._simplify(x, 0, new WeakSet());
+    },
+
+    _simplify: function(x, depth, cacheSet) {
+      if (cacheSet.has(x)) {
+        return '(cycle)';
+      }
+      if (typeof x === 'number') {
+        return x;
+      } else if (typeof x === 'string') {
+        return x.slice(0, this.maxStringLength);
+      } else if (x && x.BYTES_PER_ELEMENT) {
+        // TypedArray
+        return x.slice(0, this.maxArrayLength);
+      } else if (Array.isArray(x)) {
+        if (depth < this.maxDepth) {
+          return x.slice(0, this.maxArrayLength)
+            .map((element) => this._simplify(element, depth + 1, cacheSet));
+        } else {
+          return '[Array length=' + x.length + ']';
+        }
+      } else if (x && typeof x === 'object') {
+        cacheSet.add(x);
+        if (!isPlainObject(x)) {
+          if (x.toJSON) {
+            return this._simplify(x.toJSON(), depth, cacheSet);
+          } else if (x.toString) {
+            return this._simplify(x.toString(), depth, cacheSet);
+          } else {
+            return '(?)';
+          }
+        } else {
+          if (depth < this.maxDepth) {
+            var retObj = {};
+            var idx = 0;
+            for (var key in x) {
+              if (idx > this.maxObjectLength) {
+                break;
+              }
+              retObj[key] = this._simplify(x[key], depth + 1, cacheSet);
+              idx++;
+            }
+            return retObj;
+          } else if (x.toString) {
+            return this._simplify(x.toString(), depth, cacheSet);
+          } else {
+            return '(object?)';
+          }
+        }
+      } else if (typeof x === 'function') {
+        return '(function)';
+      } else {
+        return x;
+      }
+    }
+  }
+
+  function LogicEvent(scope, type, details) {
+    if (!(scope instanceof Scope)) {
+      throw new Error('Invalid "scope" passed to LogicEvent(); ' +
+                      'did you remember to call logic.defineScope()?');
+    }
+
+    this.scope = scope;
+    this.type = type;
+    this.details = details;
+    this.time = Date.now();
+    this.id = logic.uniqueId();
+  }
+
+  LogicEvent.fromJSON = function(data) {
+    var event = new LogicEvent(new Scope(data.namespace),
+                               data.type,
+                               data.details);
+    event.time = data.time;
+    event.id = data.id;
+    return event;
+  }
+
+  LogicEvent.prototype = {
+    get namespace() {
+      return this.scope.namespace;
+    },
+
+    toJSON: function() {
+      var event = {
+        namespace: this.scope.namespace,
+        type: this.type,
+        details: new ObjectSimplifier().simplify(this.details),
```

Comment:
In https://bugzil.la/933305 we discuss making the old logging do its JSON-conversion at the time of logging rather than the time of log dumping to contend with mutable state leading to confusion.  While it's arguably the consumer of the LogicEvents that is responsible for this, it's also arguable that for the logs to be useful they need to always be snapshotted at logging time.  `GelamTest` does the right thing since its handleEvent method which is hooked-up `logic.on('*', handleEvent);` via does `this._logs.push(event.toJSON());`.  In contrast, the circular buffer logic in _enableCircularLogging does not snapshot.

In the non-test cases we could always simplify and snapshot and discard the original details object.  (This nicely avoids pathological GC problems too.)  But the tests want access to the objects and not just a JSON.parse()d version of the log since tests likely want to be able to test object identity equivalence and potentially reach deep into structures, etc.

The other consideration is the overhead of calling toJSON.  If there is no logging going on, there's no point snapshotting.  As a nice simplification, that also means there's no point in actually creating the LogicEvent object.

Accordingly, I think I propose the following:
* Have logic.event() bail if (!this._events['*'] && !this._events[type]).  If no one's listening, no one cares, right?
* Have the LogicEvent constructor snapshot the simplified representation.  Have toJSON and toString use this representation.
* Fix or put a TODO in the circular logger for us to instead log the tuple of { event.time, event.toJSON() }.  The idea here is to avoid having the circular buffer entraining potentially huge swathes of memory thanks to LogicEvent.details but also avoid monkeying with the nice abstraction of LogicEvent too much (so no nuking its `details` attribute instead.

My rationale of checking the events rather than having some global flags is that a lot of the "slog" logging stuff that gets output to the console is actually desired by us.  I'll write a separate comment on that, but the point is that we in fact do want to only log some of the things during the course of normal operation, so we might as well do it that way.

(Side note: we probably will want to update evt.js to use a Map in the future, but definitely not in this commit.)